### PR TITLE
Fix problems in GeoIPv2 code

### LINKED ISSUE #71145: [CI] GeoIpCliTests classMethod failing on Windows
Seems this is a newly added test. The `LuceneTestCase` cleanup code doesn't play well here on Windows.

**Build scan:**
https://gradle-enterprise.elastic.co/s/se7yv5fe6d4am/tests/:distribution:tools:geoip-cli:test/org.elasticsearch.geoip.GeoIpCliTests/classMethod

**Reproduction line:**
`./gradlew.bat :distribution:tools:geoip-cli:test --tests "org.elasticsearch.geoip.GeoIpCliTests"`

**Applicable branches:**
master

**Reproduces locally?:**
Didn't try

**Failure history:**
https://gradle-enterprise.elastic.co/scans/tests?tests.container=org.elasticsearch.geoip.GeoIpCliTests&tests.test=classMethod

**Failure excerpt:**
```
java.io.IOException: Could not remove the following files (in the order of attempts):
   C:\Users\jenkins\workspace\elastic+elasticsearch+master+multijob-windows-compatibility\os\windows-2012-r2\distribution\tools\geoip-cli\build\testrun\test\temp\org.elasticsearch.geoip.GeoIpCliTests_D0C1D7D454917487-001\tempDir-003: java.nio.file.DirectoryNotEmptyException: C:\Users\jenkins\workspace\elastic+elasticsearch+master+multijob-windows-compatibility\os\windows-2012-r2\distribution\tools\geoip-cli\build\testrun\test\temp\org.elasticsearch.geoip.GeoIpCliTests_D0C1D7D454917487-001\tempDir-003
   C:\Users\jenkins\workspace\elastic+elasticsearch+master+multijob-windows-compatibility\os\windows-2012-r2\distribution\tools\geoip-cli\build\testrun\test\temp\org.elasticsearch.geoip.GeoIpCliTests_D0C1D7D454917487-001\tempDir-002: java.nio.file.DirectoryNotEmptyException: C:\Users\jenkins\workspace\elastic+elasticsearch+master+multijob-windows-compatibility\os\windows-2012-r2\distribution\tools\geoip-cli\build\testrun\test\temp\org.elasticsearch.geoip.GeoIpCliTests_D0C1D7D454917487-001\tempDir-002
   C:\Users\jenkins\workspace\elastic+elasticsearch+master+multijob-windows-compatibility\os\windows-2012-r2\distribution\tools\geoip-cli\build\testrun\test\temp\org.elasticsearch.geoip.GeoIpCliTests_D0C1D7D454917487-001\tempDir-003\target: java.nio.file.AccessDeniedException: C:\Users\jenkins\workspace\elastic+elasticsearch+master+multijob-windows-compatibility\os\windows-2012-r2\distribution\tools\geoip-cli\build\testrun\test\temp\org.elasticsearch.geoip.GeoIpCliTests_D0C1D7D454917487-001\tempDir-003\target
   C:\Users\jenkins\workspace\elastic+elasticsearch+master+multijob-windows-compatibility\os\windows-2012-r2\distribution\tools\geoip-cli\build\testrun\test\temp\org.elasticsearch.geoip.GeoIpCliTests_D0C1D7D454917487-001\tempDir-002\target: java.nio.file.AccessDeniedException: C:\Users\jenkins\workspace\elastic+elasticsearch+master+multijob-windows-compatibility\os\windows-2012-r2\distribution\tools\geoip-cli\build\testrun\test\temp\org.elasticsearch.geoip.GeoIpCliTests_D0C1D7D454917487-001\tempDir-002\target
   C:\Users\jenkins\workspace\elastic+elasticsearch+master+multijob-windows-compatibility\os\windows-2012-r2\distribution\tools\geoip-cli\build\testrun\test\temp\org.elasticsearch.geoip.GeoIpCliTests_D0C1D7D454917487-001\tempDir-002\source: java.nio.file.AccessDeniedException: C:\Users\jenkins\workspace\elastic+elasticsearch+master+multijob-windows-compatibility\os\windows-2012-r2\distribution\tools\geoip-cli\build\testrun\test\temp\org.elasticsearch.geoip.GeoIpCliTests_D0C1D7D454917487-001\tempDir-002\source
   C:\Users\jenkins\workspace\elastic+elasticsearch+master+multijob-windows-compatibility\os\windows-2012-r2\distribution\tools\geoip-cli\build\testrun\test\temp\org.elasticsearch.geoip.GeoIpCliTests_D0C1D7D454917487-001: java.nio.file.DirectoryNotEmptyException: C:\Users\jenkins\workspace\elastic+elasticsearch+master+multijob-windows-compatibility\os\windows-2012-r2\distribution\tools\geoip-cli\build\testrun\test\temp\org.elasticsearch.geoip.GeoIpCliTests_D0C1D7D454917487-001

  at __randomizedtesting.SeedInfo.seed([D0C1D7D454917487]:0)
  at org.apache.lucene.util.IOUtils.rm(IOUtils.java:319)
  at org.apache.lucene.util.TestRuleTemporaryFilesCleanup.afterAlways(TestRuleTemporaryFilesCleanup.java:216)
  at com.carrotsearch.randomizedtesting.rules.TestRuleAdapter$1.afterAlways(TestRuleAdapter.java:31)
  at com.carrotsearch.randomizedtesting.rules.StatementAdapter.evaluate(StatementAdapter.java:43)
  at com.carrotsearch.randomizedtesting.rules.StatementAdapter.evaluate(StatementAdapter.java:36)
  at org.apache.lucene.util.TestRuleAssertionsRequired$1.evaluate(TestRuleAssertionsRequired.java:53)
  at org.apache.lucene.util.TestRuleMarkFailure$1.evaluate(TestRuleMarkFailure.java:47)
  at org.apache.lucene.util.TestRuleIgnoreAfterMaxFailures$1.evaluate(TestRuleIgnoreAfterMaxFailures.java:64)
  at org.apache.lucene.util.TestRuleIgnoreTestSuites$1.evaluate(TestRuleIgnoreTestSuites.java:54)
  at com.carrotsearch.randomizedtesting.rules.StatementAdapter.evaluate(StatementAdapter.java:36)
  at com.carrotsearch.randomizedtesting.ThreadLeakControl$StatementRunner.run(ThreadLeakControl.java:375)
  at com.carrotsearch.randomizedtesting.ThreadLeakControl.lambda$forkTimeoutingTask$0(ThreadLeakControl.java:831)
  at java.lang.Thread.run(Thread.java:834)

```

### PR DESCRIPTION
**Pull request by probakowski**
This change fixes number of problems in GeoIPv2 code:

- closes streams from `Files.list` in `GeoIpCli`, which should fix tests on Windows
- makes sure that total download time in GeoIP stats is non-negative (we serialize it as `vInt` which can cause problems with negative numbers and it can happen when clock was changed during operation)
- fixes handling of failed/simultaneous downloads, #69951 was meant as a way to prevent 2 persistent tasks to index chunks but it would prevent any update if single download failed mid indexing, this change uses timestamp (`lastUpdate`) as sort of UUID. This should still prevent 2 tasks to step on each other toes (overwriting chunks) but in the end still only single task should be able to update task state (this is handled by persistent tasks framework)

Closes https://github.com/elastic/elasticsearch/issues/71145

### CODE REVIEW

**Comment by probakowski** [comment]
@elasticmachine update branch

File: modules/ingest-geoip/src/main/java/org/elasticsearch/ingest/geoip/GeoIpDownloader.java [hunk 1]
```diff
@@ -180,13 +179,13 @@ void updateTaskState() {
     }
 
     //visible for testing
-    int indexChunks(String name, InputStream is, int chunk, String expectedMd5) throws IOException {
+    int indexChunks(String name, InputStream is, int chunk, String expectedMd5, long timestamp) throws IOException {
         MessageDigest md = MessageDigests.md5();
         for (byte[] buf = getChunk(is); buf.length != 0; buf = getChunk(is)) {
             md.update(buf);
-            client.prepareIndex(DATABASES_INDEX).setId(name + "_" + chunk)
```
Comment (martijnvg) [inline modules/ingest-geoip/src/main/java/org/elasticsearch/ingest/geoip/GeoIpDownloader.java:187]: Maybe keep the _id but with timestamp? That way the _id has meaning and if due to some issue we index a document with the same _id then we fail with an error (b/c create=true).

**Review by martijnvg — APPROVED** [review]
I left two small comments. Otherwise LGTM

File: modules/ingest-geoip/src/main/java/org/elasticsearch/ingest/geoip/GeoIpDownloader.java (hunk 1 above)
Comment (probakowski) [reply modules/ingest-geoip/src/main/java/org/elasticsearch/ingest/geoip/GeoIpDownloader.java:187]: I switched back to using _id with added timestamp as you suggested

File: modules/ingest-geoip/src/main/java/org/elasticsearch/ingest/geoip/IngestGeoIpPlugin.java [hunk 2]
```diff
@@ -219,6 +219,9 @@ private static XContentBuilder mappings() {
                 .startObject("chunk")
                 .field("type", "integer")
                 .endObject()
+                .startObject("timestamp")
+                .field("type", "long")
```
[the remark this answers is not in the document]
Comment (probakowski) [reply modules/ingest-geoip/src/main/java/org/elasticsearch/ingest/geoip/IngestGeoIpPlugin.java]: as timestamp is part of the id now it doesn't need to be indexed separately. I've removed it from mapping

**Pull request merged**